pex/fetcher.py

```diff
@@ -50,11 +56,30 @@ def __init__(
         handlers = [ProxyHandler(proxies), HTTPSHandler(context=ssl_context)]
         if handle_file_urls:
             handlers.append(FileHandler())
+
+        self._password_database = PasswordDatabase.from_netrc(netrc_file=netrc_file).append(
+            password_entries
+        )
         self._handlers = tuple(handlers)
 
     @contextmanager
     def get_body_stream(self, url):
         # type: (Text) -> Iterator[BinaryIO]
+
+        handlers = list(self._handlers)
+        if self._password_database.entries:
+            password_manager = HTTPPasswordMgrWithDefaultRealm()
```

Comment:
.netrc entries are keyed by machine name (without port) but are expected to be used for all auth to that machine, on any port. Or at least this is how curl behaves. 

So at first glance this seems broken, because the uri key we pass to add_password() doesn't include a port. We could inject the port in `uri_or_default()` since we know it there. But, from my reading of the urllib2 source code it looks like we don't need to, because that code strips any port information on both sides before comparing entries.

So I guess this comment is to confirm that this is also your understanding, and to say that perhaps this complication should be documented.

Reply:
The urllib2 code also appears to ignore scheme information and will send a password associated with an https URL to an http URL on the same host. But 🤷

Reply:
> So at first glance this seems broken, because the uri key we pass to add_password() doesn't include a port. We could inject the port in uri_or_default() since we know it there.

It does include a port if the netrc entry happens to have a port. If it doesn't have a port (a traditional netrc entry), urllib actually creates a second entry that does include the port (80 or 443 as appropriate).

```
$ python2.7
Python 2.7.18 (default, Oct 10 2021, 22:29:32) 
[GCC 11.1.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from urllib2 import HTTPPasswordMgrWithDefaultRealm
>>> manager = HTTPPasswordMgrWithDefaultRealm()
>>> manager.add_password(realm=None, uri="http://foo:9999", user="foo", passwd="bar")
>>> manager.add_password(realm=None, uri="http://bar", user="bar", passwd="baz")
>>> manager.passwd
{None: {(('foo:9999', '/'),): ('foo', 'bar'), (('bar', '/'),): ('bar', 'baz'), (('bar:80', '/'),): ('bar', 'baz')}}
```

IOW it's the opposite to what you suggest, urllib compares against the host name (possibly containing port) - it does not strip port.

I'll add some comment.